### PR TITLE
Initial policy code

### PR DESCRIPTION
**Pull request by jvanz**
## Description

Adds the initial code to perform the Namespace label propagation.

Fix https://github.com/kubewarden/kubewarden-controller/issues/452

### CODE REVIEW

File: questions-ui.yml
Comment (jvanz) [inline questions-ui.yml:1]: @jordojordo can you review this file? :)

File: test_data/pod.json
Comment (flavio) [inline test_data/pod.json:3]: wrong

File: test_data/pod_with_no_labels.json
Comment (flavio) [inline test_data/pod_with_no_labels.json:3]: wrong

File: test_data/pod_with_wrong_label_value.json
Comment (flavio) [inline test_data/pod_with_wrong_label_value.json:3]: wrong

File: validate.go
Comment (flavio) [inline validate.go:4]: leftover, to be removed

File: validate.go
Comment (flavio) [inline validate.go:42]:
```suggestion
		return nil, fmt.Errorf("Cannot get namespace data: %s", err)
```

File: validate.go
Comment (flavio) [inline validate.go:46]:
```suggestion
		return nil, fmt.Errorf("Cannot parse namespace data: %s", err)
```

File: validate.go
Comment (flavio) [inline validate.go:137]: in addition to that, I think we should also look for the labels of the PodSpec and ensure the final Pod has the relevant labels too

File: validate.go
Comment (flavio) [inline validate.go:149]: in addition to that, I think we should also look for the labels of the PodSpec and ensure the final Pod has the relevant labels too

File: validate.go
Comment (flavio) [inline validate.go:161]: in addition to that, I think we should also look for the labels of the PodSpec and ensure the final Pod has the relevant labels too

File: validate.go
Comment (flavio) [inline validate.go:173]: in addition to that, I think we should also look for the labels of the PodSpec and ensure the final Pod has the relevant labels too

File: validate.go
Comment (flavio) [inline validate.go:185]: in addition to that, I think we should also look for the labels of the PodSpec and ensure the final Pod has the relevant labels too

File: validate.go
Comment (flavio) [inline validate.go:197]: in addition to that, I think we should also look for the labels of the PodSpec and ensure the final Pod has the relevant labels too

File: validate.go
Comment (flavio) [inline validate.go:209]: in addition to that, I think we should also look for the labels of the PodSpec and ensure the final Pod has the relevant labels too

File: validate.go
Comment (flavio) [inline validate.go:52]: nitpicking, I think `labelsToPropagate` is a better name

File: validate.go
Comment (flavio) [inline validate.go:121]: simpler, more direct
```suggestion
			has_mutation = true
```

File: validate.go
Comment (flavio) [inline validate.go:116]: nitpicking:
* `labels` -> `currentLabels`
* `newLabels` -> `labelsToPropagate`

File: validate_test.go
Comment (flavio) [inline validate_test.go:71]: nitpicking: `s/ok/found/

**Review by flavio — CHANGES_REQUESTED** [review]
Overall ok, I left some comments. Thanks for having implemented this policy!

File: e2e.bats
Comment (viccuad) [inline e2e.bats:11]: I think we need to amend or remove all "request rejected" comments

File: e2e.bats
Comment (viccuad) [inline e2e.bats:14]: It would be better to grep for the new label value to check that it was changed.

**Review by viccuad — COMMENTED** [review]
Overall good modulo comments from Flavio!

**Review by jordojordo — APPROVED** [review]
`questions-ui.yml` looks good!

File: go.sum
Comment (jvanz) [reply go.sum:12]: Yes. I've kept the old version because I did not want to wait for the SDK release. But as we already did, I'll update it.

File: validate_test.go
Comment (jvanz) [reply validate_test.go:1]: I'm working on that.

File: validate.go
Comment (flavio) [inline validate.go:115]: nitpicking: `currentLabelsToPropagate` -> `labelsToPropagate`

Comment (flavio) [reply validate.go:115]: Also, it would be nice to add some inline docs explaining the function signature, especially what the `bool` return value is about

File: validate.go
Comment (flavio) [inline validate.go:60]: This function is used only by the `getNamespace` function, the purpose is to extract the object metadata from the request, so that we can understand which namespace is owning the resource.

We can obtain this information in a faster way: the `AdmissionRequest` has a `Namespace` attribute that already holds this information. See [here](https://pkg.go.dev/github.com/kubewarden/policy-sdk-go@v0.4.1/protocol#KubernetesAdmissionRequest).

I propose to use this attribute and just remove this function entirely

File: validate.go
Comment (flavio) [inline validate.go:143]: It's a matter of personal taste, but I don't like this line. I would prefer something that is more straightforward and explicit:

```golang
obj_meta_mutated := updateLabels(&deployment.Metadata.Labels, labelsToPropagate)
spec_template_mutated := updateLabels(&deployment.Spec.Template.Metadata.Labels, labelsToPropagate)
if obj_meta_mutated || spec_template_mutated {
	return kubewarden.MutateRequest(deployment)
}
```
The same applies to the other snippets of code that looks like that.

Speaking of which, there's quite some repetition here:
  * check the main object metadata.labels, create a map if needed
  * check the template.metadata.labels, create a map if needed
  * check if some mutation was done, produce the final response

What about doing something like that:

```golang
switch object.Request.Kind.Kind {
case REPLICASET_KIND:
   replicaset := appsv1.ReplicaSet{}
   // unmarshal
   obj_meta_mutated := propagateLabels(&replicaset.Metadata, labelsToPropagate)
   spec_template_mutated := propagateLabels(&replicaset.Spec.Template.Metadata, labelsToPropagate)
   if obj_meta_mutated || spec_template_mutated {
      return mutate
   }
case POD_KIND:
   // similar to above, it just doesn't have the spec_template_mutated
   if mutated {
      return mutate
   }
}  
// if we got there it means no mutation happened
return accept()
```

The `propagateLabels` function would be just `updateLabels` renamed, with some extra code:

```golang
func propagateLabels(meta *Metadata, labelsToPropagate map[string]string) bool {
  if meta.Labels == nil {
     meta.Labels = make(map[string]string)
  }
  // the code of updateLabels
}
```

BTW: I just realized I'm too used to Rust, I named a bunch of variables like Rust expects :sweat_smile:

File: validate.go
Comment (jvanz) [reply validate.go:115]: Ops, I've caused rename issue after a vim find and replace command. Sorry.

File: validate.go
Comment (flavio) [inline validate.go:33]: Warning: if the user enforced with policy targeting a cluster-wide resource, `validation.Request.Namespace` might be empty or nil (I don't know what is the default for this go struct). Please a check for that and return an error

File: validate.go
Comment (flavio) [inline validate.go:73]: nit: empty line

File: README.md
Comment (flavio) [inline README.md:11]: Documenting the new feature that you implemented:

```suggestion
This policy is able to set the labels for the following resource kinds: `Pod`,
`ReplicationController`, `Deployment`, `ReplicaSet`, `StatefulSet`, `DaemonSet`, `Job` and `CronJob`.

When dealing with Kubernetes resources that generate pods, the policy ensures the special labels
are propagated also to them.
```

Once merged, don't forget to update the artifact hub metadata.

**Review by flavio — CHANGES_REQUESTED** [review]
Looks good, one last thing to update

**Pull request merged**
